go-runtime/compile/build.go

```diff
@@ -262,17 +271,80 @@ func buildDir(moduleDir string) string {
 	return filepath.Join(moduleDir, buildDirName)
 }
 
+// OngoingState maintains state between builds, allowing the Build function to skip steps if nothing has changed.
+type OngoingState struct {
+	imports   []string
+	moduleCtx mainModuleContext
+}
+
+func (s *OngoingState) checkIfImportsChanged(imports []string) (changed bool) {
+	if slices.Equal(s.imports, imports) {
+		return false
+	}
+	s.imports = imports
+	return true
+}
+
+func (s *OngoingState) checkIfMainModuleContextChanged(moduleCtx mainModuleContext) (changed bool) {
+	if stdreflect.DeepEqual(s.moduleCtx, moduleCtx) {
+		return false
+	}
+	s.moduleCtx = moduleCtx
+	return true
+}
+
+// DetectedFileChanges should be called whenever file changes are detected outside of the Build() function.
+// This allows the OngoingState to detect if files need to be reprocessed.
+func (s *OngoingState) DetectedFileChanges(config moduleconfig.AbsModuleConfig, changes []watch.FileChange) {
+	paths := []string{
+		filepath.Join(config.Dir, ftlTypesFilename),
+		filepath.Join(config.Dir, "go.mod"),
+		filepath.Join(config.Dir, "go.sum"),
+	}
+	for _, change := range changes {
+		if !slices.Contains(paths, change.Path) {
+			continue
+		}
+		// If files altered by Build() have been manually changed, reset state to make sure we correct them if needed.
+		s.reset()
+		return
+	}
+}
+
+func (s *OngoingState) reset() {
+	s.imports = nil
+	s.moduleCtx = mainModuleContext{}
+}
+
 // Build the given module.
-func Build(ctx context.Context, projectRootDir, stubsRoot string, config moduleconfig.AbsModuleConfig, sch *schema.Schema, filesTransaction watch.ModifyFilesTransaction, buildEnv []string, devMode bool) (moduleSch optional.Option[*schema.Module], buildErrors []builderrors.Error, err error) {
+func Build(ctx context.Context, projectRootDir, stubsRoot string, config moduleconfig.AbsModuleConfig,
+	sch *schema.Schema, deps, buildEnv []string, filesTransaction watch.ModifyFilesTransaction, ongoingState *OngoingState,
+	devMode bool) (moduleSch optional.Option[*schema.Module], buildErrors []builderrors.Error, err error) {
 	if err := filesTransaction.Begin(); err != nil {
 		return moduleSch, nil, fmt.Errorf("could not start a file transaction: %w", err)
 	}
 	defer func() {
 		if terr := filesTransaction.End(); terr != nil {
 			err = fmt.Errorf("failed to end file transaction: %w", terr)
 		}
+		if err != nil {
```

Comment:
Shouldn't we log this error?

Reply:
All errors returned from this function get logged by the build engine, so we don't need to double log here.